New Exercise: Scrabble Score

jvarness opened the pull request:
I haven't actually ran the presubmit and formatter check on this code yet. Just getting this out there for review.

jvarness commented:
Hey, what's the best way to get the `config.json` generated for this?

Stargator commented:
Hey @jvarness, thanks for coming over from dev.to!

[CONTRIBUTING](https://github.com/exercism/dart/blob/master/CONTRIBUTING.md) should have the relevant guidance.

I don't believe we re-generate the config.json. You would just need to update it with the information for the new exercise. Preferably at the bottom of the list. Guess for any values you may not be sure about and we can review them later.

Stargator commented:
@jvarness In the README it suggests to use dart bin/create_exercise.dart to generate the test file and some related files.

You can run it with `dart bin/create_exercise.dart`.

Do comment out the following line from it so it overwrites your existing files and you can diff the changes and make any corrections you feel warranted.

```dart
  if (await exerciseDir.exists()) {
    stderr.write("$name already exist\n");
    exit(1);
  }
```

Stargator commented:
> @Stargator I tried executing bin/create_exercise.dart, but all it did was get rid of all the work I have already done.

Do you have a solid Diff tool? Not sure what platform you're using, but after you run the script, use a Diff comparison tool like BeyondCompare to see the differences. Then you can manually make some corrections.

Stuff like adding the `skip` automatically and consistently generation of the README and other files would help. If you feel uncomfortable doing that, then I'll do it locally and just add comments for the suggested changes.

Stargator commented on exercises/scrabble-score/README.md:
```diff
@@ -0,0 +1,45 @@
+- You can play a double or a triple letter.
+- You can play a double or a triple word.
+
+## Source
```
Add the following above Source, including a line before and after:

```Markdown

To run the tests:

```sh
$ pub run test
```

For more detailed info about the Dart track see the [help page](http://exercism.io/languages/dart).

```

Stargator commented on exercises/scrabble-score/README.md:
```diff
@@ -0,0 +1,45 @@
+Given a word, compute the scrabble score for that word.
```
Add the following as the first line, then add another line after it:

```Markdown
# Scrabble Score

```

jvarness replied:
That's been added.

Stargator commented on exercises/scrabble-score/lib/scrabble_score.dart:
```diff
@@ -0,0 +1,3 @@
+class ScrabbleScore {
+  // your code here
```
Change to `// Put your code here`

Add a line after the `}`

jvarness replied:
Added

jvarness replied:
Sorry. Added `Put`

Stargator commented on exercises/scrabble-score/test/scrabble_score_test.dart:
```diff
@@ -0,0 +1,168 @@
+      test('strings containing only whitespace characters', () {
+        var word = new ScrabbleScore(' \t\r\n');
+        expect(word.getScrabbleScore(), equals(0));
+      });
```
Add `skip: true` to all tests after the first.

Add `skip: false` to the first test.

jvarness commented:
@Stargator I'm going to probably end up rewriting the test cases to more closely match the canonical-data for the problem. I was trying to be cool and write all my own.

This is the first exercise I've tried to implement. Next time, I will use the exercise generator so that everything will be less painful to review.

jvarness commented:
@Stargator I've updated the test cases to only include those specified in the canonical data.

Stargator commented:
@jvarness No worries, thanks for adding a new exercise. It helps so i understand where some of the pain points might be for new contributors.

Stargator commented on exercises/scrabble-score/test/scrabble_score_test.dart:
```diff
@@ -0,0 +1,72 @@
+
+      test('lowercase letters', () {
+        var word = new ScrabbleScore('a');
+        expect(word.getScrabbleScore(), equals(1));
```
Just noting that to fit the canonical-data.json's template the tests should look more like:

```dart
    test("lowercase letters", () {
      final int result = scrabbleScore.score("a");
      expect(result, equals(1));
    });
```

No constructor, just a method called score that is passed the input.

jvarness replied:
If that's the case, would it be more proper to utilize Dart's support for high-order functions? No classes, just functions?

jvarness commented:
@Stargator I've updated the example and the test cases and example to support high-order functions. I've also rebased off of the current master and squashed my commits.

Stargator commented on exercises/scrabble-score/lib/example.dart:
```diff
@@ -0,0 +1,39 @@
+  'q': 10
+};
+
+int getScrabbleScore(String word) {
```
Can you rename the function to score? So it fits the property "property" that's defined in it.

jvarness replied:
Done

Stargator commented on exercises/scrabble-score/lib/example.dart:
```diff
@@ -0,0 +1,39 @@
+
+int getScrabbleScore(String word) {
+  int val = 0;
+  if(word.length > 0) {
```
Run `dartfmt -i 0 -l 120 -w .` so the exercise is properly formatted.

jvarness replied:
Done

Stargator commented on exercises/scrabble-score/pubspec.yaml:
```diff
@@ -0,0 +1,3 @@
+name: 'scrabble_score'
+dev_dependencies:
+  test: '<0.13.0'
```
In the create_exercise_test.dart, extend the timeout to 2 minutes. As the tests are timing out on Travis CI.

Though the builds are not failing properly, so I'm going to make another ticket for it. So I'll have to double check the contents to ensure the results are correct.

jvarness replied:
I think you may have meant `exercises_test.dart`?

I changed the Duration to 120 seconds.

Stargator replied:
I stand corrected, yep.

jvarness commented:
@Stargator I believe I've covered everything.

Stargator commented:
@jvarness Sorry for the back and forth, I thought I had already submitted the feedback about the null.

Thanks for your work and patience.

jvarness commented:
@Stargator no problem. I changed the `config.json` file to have `null` and not `"null"`

Stargator commented:
@jvarness Reviewing today, just gotta finish refactoring something on another project and then I'll run through your changes.

Stargator commented:
Approved, just recommend moving:
>     {
>      "core": true,
>      "difficulty": 1,
>      "slug": "scrabble-score",
>      "topics": [
>        "control_flow_conditionals",
>        "control_flow_loops",
>        "equality",
>        "strings"
>      ],
>      "unlocked_by": null,
>      "uuid": "d74be875-bc1c-47a5-bf80-0c4c57ed200e"
>    }

To just about the entry for armstrong numbers

jvarness commented:
I've moved the exercise above the armstrong numbers exercise. Let me know if there's anything else you'd like for me to do.
